OpenRA.Mods.Cnc/Traits/SupportPowers/DropPodsPower.cs

```diff
@@ -0,0 +1,133 @@
+using OpenRA.Mods.Common;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	public class DropPodsPowerInfo : SupportPowerInfo, IRulesetLoaded
+	{
+		[FieldLoader.Require]
+		[Desc("Drop pod unit")]
+		[ActorReference(typeof(AircraftInfo), typeof(FallsToEarthInfo))]
+		public readonly string[] UnitTypes = null;
+
+		[Desc("Number of drop pods spawned.")]
+		public readonly int2 Drops = new int2(5, 8);
+
+		[Desc("Sets the approach direction.")]
+		public readonly int PodFacing = 32;
+
+		[Desc("Maximum offset from targetLocation")]
+		public readonly int PodScatter = 3;
+
+		[Desc("Effect sequence sprite image")]
+		public readonly string Effect = "explosion";
+
+		[Desc("Effect sequence to display")]
+		[SequenceReference("Effect")]
+		public readonly string EffectSequence = "piffs";
+
+		[PaletteReference]
+		public readonly string EffectPalette = "effect";
+
+		[Desc("Which weapon to fire")]
+		[WeaponReference]
+		public readonly string Weapon = "Vulcan2";
+
+		public WeaponInfo WeaponInfo { get; private set; }
+
+		[Desc("Apply the weapon impact this many ticks into the effect")]
+		public readonly int WeaponDelay = 0;
+
+		public override object Create(ActorInitializer init) { return new DropPodsPower(init.Self, this); }
+
+		public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			WeaponInfo weapon;
+			var weaponToLower = (Weapon ?? string.Empty).ToLowerInvariant();
+			if (!rules.Weapons.TryGetValue(weaponToLower, out weapon))
+				throw new YamlException("Weapons Ruleset does not contain an entry '{0}'".F(weaponToLower));
+
+			WeaponInfo = weapon;
+
+			base.RulesetLoaded(rules, ai);
+		}
+	}
+
+	public class DropPodsPower : SupportPower
+	{
+		readonly DropPodsPowerInfo info;
+
+		public DropPodsPower(Actor self, DropPodsPowerInfo info)
+			: base(self, info)
+		{
+			this.info = info;
+		}
+
+		public override void Activate(Actor self, Order order, SupportPowerManager manager)
+		{
+			base.Activate(self, order, manager);
+
+			SendDropPods(self, order, info.PodFacing);
+		}
+
+		public void SendDropPods(Actor self, Order order, int podFacing)
+		{
+			var actorInfo = self.World.Map.Rules.Actors[info.UnitTypes.First().ToLowerInvariant()];
+			var aircraftInfo = actorInfo.TraitInfo<AircraftInfo>();
+			var altitude = aircraftInfo.CruiseAltitude.Length;
+			var approachRotation = WRot.FromFacing(podFacing);
+			var fallsToEarthInfo = actorInfo.TraitInfo<FallsToEarthInfo>();
+			var delta = new WVec(0, -altitude * aircraftInfo.Speed / fallsToEarthInfo.Velocity.Length, 0).Rotate(approachRotation);
+
+			self.World.AddFrameEndTask(w =>
+			{
+				PlayLaunchSounds();
+
+				var target = order.Target.CenterPosition;
+				var targetCell = self.World.Map.CellContaining(target);
+				var podLocations = self.World.Map.FindTilesInCircle(targetCell, info.PodScatter)
+					.Where(c => aircraftInfo.LandableTerrainTypes.Contains(w.Map.GetTerrainInfo(c).Type));
```

Comment:
It is now possible to have zero drop-pods actually spawning when you activate the ability over a dense crowd. (and yet the guns still fire, which also looks odd.) It would be better to add the `CanLand` check here together with the terrain types check instead of disposing after a `podLocation` is already selected.

Reply:
How did the original game handle this situation? this should be simple enough to test and copy.

Reply:
Moved the `DropPodImpact` after the pod creation and `CanLand` checks.

Reply:
I did some mucking around with drop pods in the original and I wasn't able to create a crowd big enough to block all the pods before giving up. It is a bit difficult to say whether or not any pods where blocked, since the total number of pods is random, but I suspect that the pods will always choose an unoccupied cell no matter how far away it must go to find it. The pods do always cluster around each other though; I have never seen them end up at different ends of the crowd. I also noticed that when I left a single unoccupied square in the middle of the crowd, multiple drop pods will land in the same cell.

Moving `DropPodImpact` solves the issue of the guns firing out of nowhere, but it doesn't prevent the pods from choosing an occupied cell at random (when other unoccupied cells are available) and getting needlessly blocked.

Reply:
Added additional checks to combat that.